packages/suite/src/components/suite/NumberInput.tsx

```diff
@@ -0,0 +1,227 @@
+            }
+
+            const formattedNumber = localizeNumber(cleanValueString(value, locale), locale);
+            return handleSetDisplayValue(formattedNumber);
+        },
+        [ref, locale],
+    );
+
+    useLayoutEffect(() => {
+        if (
+            value &&
+            previousFormValue.current &&
+            cleanValueString(previousFormValue.current, locale) !==
+                cleanValueString(value, locale) &&
+            cleanValueString(previousDisplayValue.current, locale) ===
+                cleanValueString(displayValue, locale)
+        ) {
+            formatDisplayValue(value);
+        }
+    }, [previousFormValue, previousDisplayValue, formatDisplayValue, displayValue, value, locale]);
+
+    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        const { value } = e.target;
+        let inputValue = value;
+
+        // make the string compliant with Number
+        const cleanInput = cleanValueString(inputValue, locale);
+
+        const { thousandsSeparatorCharCode } = getLocaleSeparatorCharCodes(locale);
+        // allow inputs like '.031' or ',1' and disallow anything non-numerical
+        if (
+            !DECIMAL_SEPARATOR_CODES.includes(inputValue.charCodeAt(0)) &&
+            Number.isNaN(Number(cleanInput))
+        ) {
+            return;
+        }
+
+        // read cursor position before formatting
+        let { selectionStart } = ref.current;
+
+        if (inputValue.length < value.length) {
+            // handle deleting a thousands separator for a non-empty input,
+            // otherwise the number instantly gets formatted back to having that separator
+            const deletedCharacterCode = displayValue[selectionStart]?.charCodeAt(0);
+            if (deletedCharacterCode === thousandsSeparatorCharCode && inputValue.length) {
+                inputValue =
+                    displayValue.substring(0, selectionStart - 1) +
+                    displayValue.substring(selectionStart);
+            }
+        }
+
+        // format and set display value
+        const currentValueLength = ref.current.value.length;
+        const formattedValue = formatDisplayValue(inputValue);
+        if (!formattedValue) {
+            return;
+        }
+        const formattedValueLength = formattedValue.length;
+
+        // pass cleaned value to the form
+        previousFormValue.current = cleanInput;
+        onChange(cleanInput);
+        onChangeCallback?.(cleanInput);
+
+        // detect if spaces have been added/removed
+        if (currentValueLength - formattedValueLength < 0) {
+            selectionStart++;
+        } else if (currentValueLength - formattedValueLength > 0) {
+            selectionStart--;
+        }
+
+        // make sure the cursor stays in the right place after formatting
+        ref.current.setSelectionRange(selectionStart, selectionStart);
+    };
+
+    // copy the non-formatted value
+    const handleCopy = useCallback(
```

Comment:
Yeah, since Input is not a simple HTML component and those functions are not small it makes sense IMO. Possible that they are not required, I just don't believe memoizing is bad in cases like this.